### DISCUSSION TITLE
filtering items popup

### DISCUSSION THREAD

**Discussion by yinonov** [Vonage/vivid#949] [New Developments]
# Filtering list items popup
## A common component seen across the organization
It is a popup surface box with the anatomy of a **header**, **body**, **footer**:

- ***header*** contains input to filter data list
- ***body*** reflects the filtered list items
- ***footer*** containing the action buttons

![image](https://user-images.githubusercontent.com/15140652/125295601-111c4c00-e32e-11eb-8abf-86892cc4db77.png)

the request to consolidate such a pattern to a dedicated component is challenged by its generics and is often built with an opinionated approach under the hood (pre-rendered/lazy-rendered/hidden items, virtual scroll, should it have selection state like in data grid etc'...)

## Current provision
Button triggers a dropdown with developer own content and slots for **header**, **body**, **footer**.
same result can be achieved by little extra code.
this brings up the question whether the common usage raises the priority of having a dedicated component featuring a unified behavior. if so, this needs a spec of its input, output, events, methods, exposed properties and more...
```html
<div style="position: relative">
  <vwc-button label="Open dropdown" icon="chevron-down-line"></vwc-button>
  <vwc-dropdown>
    <vwc-textfield slot="header" placeholder="Search..." icon="search-line" dense></vwc-textfield>
    <vwc-data-grid id="data-grid">
      <vwc-data-grid-column class="selector" selector="multi"></vwc-data-grid-column>
      <vwc-data-grid-column path="label" header="Entity ID"></vwc-data-grid-column>
    </vwc-data-grid>
    <vwc-button slot="actions">Cancel</vwc-button>
    <vwc-button slot="actions" layout="filled">Done</vwc-button>
  </vwc-dropdown>
</div>
```

##### GitHub
![image](https://user-images.githubusercontent.com/15140652/125295284-be429480-e32d-11eb-8087-ea804d846a36.png)

## Considered Alternatives

- [Combobox](https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.1pattern/listbox-combo.html) - similar to [datalist](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist) but with multiple selection
![image](https://user-images.githubusercontent.com/15140652/125296621-25ad1400-e32f-11eb-8d9f-f2ebf75b6c6d.png)

**Comment by rsreeju** [comment]
From a perspective of an autocomplete, we are looking for something which is more intuitive(like a textbox) and does not involve a button to open the dropdown, The above is a data-grid encapsulated in a modal, and there are too many actions the user needs to perform to select something, An autocomplete should be pretty intuitive like the combobox mentioned above (but the example 1 or 2 in this-> https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.1pattern/listbox-combo.html) without the dropdown button

**Comment by adam-trojanowski** [comment]
The main problem with having it simply as a list of building blocks is consistency. Given that this is going to be used across different products and technologies it would be beneficial to limit the possibilities. Hide some of the logic needed for this kind of component to work.

If we leave it as is each time someone builds something similar they will have to answer the same questions:
- what are the size restrictions,
- how to properly position it and what kind of trigger element to use,
- how should the selection behave - only a checkbox or additional styles for rows,
- should the dropdown close on click outside,
- how to show if no elements match the search,

They might not be big issues alone but might cause a lot of confusion when implemented in different ways in different places. We created such a component in our application and had to decide on the behavior. It would be a lot easier if we had the possibility to use a component that works the way the UI and UX designers want.

Also, there might be incompatibilities between different components. Like e.g. we used a list component that was clearing the selected items each time we filtered the elements. Or not being able to use a select component as a trigger.

**Comment by yinonov** [comment]
will be handled in https://jira.vonage.com/browse/VIV-654

**Comment by yinonov** [comment]
[related request by UP](https://vonage.slack.com/archives/C013F0YKH99/p1629202147066500) require virtual scrolling to handle large length items

**Comment by ydinari** [comment]
To me it seems that this should just be the select component, with extra optional parts like the search box, header/footer etc.
A feature not mentioned here that might be useful is some way to divide the items into categories (the example I have in mind is extensions categories in VBC, like call-group, call-queue, regular extension, VR, etc.).
There should be a way to see for each item which category it is, and also search by category.
Another thing which I assume is included but didn't see mentioned is that it must have a virtual scroll, o/w it will be useless for large amounts of data.

**Comment by jshenkman** [comment]
Another use case requested from AI


![Screen Shot 2021-08-18 at 15 40 51](https://user-images.githubusercontent.com/67224525/129899668-62d7d7d2-aff5-49a2-b485-44504ce2d2d0.png)


Adding the selected items from the dropdown to a chip-set in the input field.

**Comment by yinonov** [comment]
![image](https://user-images.githubusercontent.com/15140652/140028426-d9edb7f9-b3e8-405d-a8ec-53a17fbd0451.png)
(proposed components architecture)

**Comment by jshenkman** [comment]
![Screen Shot 2021-11-10 at 14 19 40](https://user-images.githubusercontent.com/67224525/141112021-ce0e6d13-b034-49bc-938a-96e6e62506ac.png)

**Comment by yinonov** [comment]
Follow https://www.w3.org/TR/wai-aria-practices-1.2/examples/combobox/combobox-select-only.html
